### PR TITLE
tests: add new threads permissions tests, modify existing

### PR DESCRIPTION
**Pull request by maksimmasalski**
1. Found out that thread tests doesn't test next ideas of requirements,
which I think necessary to be tested and verified:
-the kernel need to prevent user threads creating new threads from
using thread or thread stack objects which are in an initialized state
-Upon thread exit, the kernel need to mark the exiting thread
and thread stack objects as uninitialized

Add new tests to test requirements above, that way we can cover more
features to be tested:
- test_new_user_thread_with_in_use_stack_obj()
- test_mark_thread_exit_uninitialized()

2. Modified test test_create_new_thread_from_user() to verify that kernel
provides new user threads access to their own thread object.

3. Also I added detailed Doxygen tags for each new test and existing
modified test.
4. Added Doxygen tag to the existing test test_stack_buffer, it
covers requirement:
-The kernel need to provide all threads read and write access to their
own stack memory buffer.

Signed-off-by: Maksim Masalski maksim.masalski@intel.com

### CODE REVIEW

**Comment by maksimmasalski** [comment]
@nashif @andrewboie @andyross I'm looking for code review of that PR.

**Review by andyross — APPROVED** [review]
These all look good.  @andrewboie will want to chime in on the details of the kernel object lifecycle

File: tests/kernel/mem_protect/mem_protect/src/kobject.c [hunk 1]
```diff
@@ -886,3 +944,120 @@ void test_create_new_invalid_prio_thread_from_user(void *p1, void *p2, void *p3)
...
 }
+
+/* Function to check read/write access to thread's own stack memory buffer */
+static void thread_stack_rw_access(void *p1, void *p2, void *p3)
+{
+	set_fault_valid(false);
+
+	u32_t stack_buf[0];
+	/* write stack data */
+	stack_buf[0] = 1234;
```
Comment (andrewboie) [inline tests/kernel/mem_protect/mem_protect/src/kobject.c:955]: this doesn't really test that you can read/write the entire stack buffer, just the first 4 bytes.

Given that tests/kernel/mem_protect/userspace already check this, why was this added?
```
		/* If we're in user mode, check every byte in the stack buffer
		 * to ensure that the thread has permissions on it.
		 */
		for (pos = stack_start; pos < stack_end; pos++) {
			zassert_false(check_perms((void *)pos, 1, 1),
				      "bad MPU/MMU permission on stack buffer at address %p",
				      pos);
		}
```

**Review by tyrantlizard77 — COMMENTED** [review]
the stack buffer test isn't very thorough, just testing the first few bytes, and there is already atest for this

File: tests/kernel/mem_protect/mem_protect/src/kobject.c (hunk 1 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c:955]: @andrewboie As I understand I can remove my test test_stack_user_rw_access, because it is already tested. Right?

Comment (nashif) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c]: you should squash the commits, adding a test in 1 commit and then removing it in the next has no value.

Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/kobject.c]: @nashif Updated.

**Review by nashif — APPROVED** [review]
